### DISCUSSION TITLE
「重学TS 2.0 」TS 面试题：聊聊你对 TS 类中 public、private、protected、readonly 修饰符的理解？

### DISCUSSION THREAD

**Discussion by semlinker** [semlinker/awesome-typescript#79] [General]
### 聊聊你对 TS 类中 public、private、protected、readonly 修饰符的理解？

**Comment by wenkm** [comment]
`public` 默认就这玩意儿，随便用
`private` 我自己家的，只能我家使用
`protected` 我保管的，我和我的下属才能用
`readonly` 只可远观不可亵玩焉

**Comment by atom-7** [comment]
private 只能在本类中使用, 实例化后和在实例外部无法访问到
protected 自己和子类访问可用
readonly 都可以获取,只是无法修改

**Comment by douhao1988** [comment]
`public`: 成员都默认为`public`，被此限定符修饰的成员是可以被外部访问；
`private`: 被此限定符修饰的成员是只可以被类的内部访问；
`protected`: 被此限定符修饰的成员是只可以被类的内部以及类的子类访问;
`readonly`: 关键字将属性设置为只读的。 只读属性必须在声明时或构造函数里被初始化。

**Comment by pingan8787** [comment]
按照字面意思先看下这几个词：
- public：公共的，公开的；
- private：私有的，私人的；
- protected：受保护的；
- readonly：只读的；

从这几个词的含义，大致知道这几个修饰符是做啥事的。

在 TypeScript 的类中，可以用这几个修饰符来修饰其中的属性或方法。其中：

- `public` 修饰符是默认成员属性或方法都是，可以在任何地方访问；
- `private` 修饰符修饰的属性或方法是私有的，不能在声明它的类的外部访问；
- `protected`修饰符修饰的属性或方法将受保护，仅仅只能在其子类中访问；
- `readonly`  修饰符修饰的属性只读，只允许出现在属性声明或者索引签名或者构造函数中，可以防止被修改；

文档地址：
- [public](https://www.typescriptlang.org/docs/handbook/2/classes.html#public)
- [private](https://www.typescriptlang.org/docs/handbook/2/classes.html#private)
- [protected](https://www.typescriptlang.org/docs/handbook/2/classes.html#protected)
- [readonly](https://www.typescriptlang.org/docs/handbook/2/classes.html#readonly)
